executor/distsql.go

```diff
@@ -239,10 +240,46 @@ func (e *IndexReaderExecutor) Next(ctx context.Context, req *chunk.Chunk) error
 		return nil
 	}
 
+	// Global index has the last partitionID column, filter out needed rows and prune the last column.
+	if e.index.Global {
+		req.Reset()
+		retTps := append(e.base().retFieldTypes, types.NewFieldType(mysql.TypeLonglong))
+		if err := e.result.ResetFieldTypes(retTps); err != nil {
+			return err
+		}
+
+		// TODO(L-Maple): chk should be reused
+		chk := chunk.NewChunkWithCapacity(retTps, req.Capacity())
+		err := e.result.Next(ctx, chk)
+		if err != nil {
+			e.feedback.Invalidate()
+			return err
+		}
+
+		for i := 0; i < chk.NumRows(); i++ {
+			row := chk.GetRow(i)
+			pidOffset := row.Len() - 1
+			pid := row.GetInt64(pidOffset)
+			if _, exist := e.partitionIDMap[pid]; !exist {
+				continue
+			}
+
```

Comment:
Is this the reason for the partitionIDMap, using it for pruning when reading the index, to avoid extra processing?

Could we not also push down that filtering, i.e. to avoid retrieving it from TiKV?

Reply:
You are right. I found that if we handle filter logic in TiDB, there are at least two problems: 1) performance; 2) lots of tests have to be modified, as cost is changed. 
By following your advice, I propose the tipb PR: https://github.com/pingcap/tipb/pull/297?w=1. Please review it.

Reply:
Why not filter out unnecessary pids when creating kvrange? Then we don't need change TiKV and TiPB any more.